Team417_CENTERSTAGE/src/main/java/org/firstinspires/ftc/team417_CENTERSTAGE/baseprograms/BaseTeleOp.java

```diff
@@ -0,0 +1,210 @@
+    public MecanumDrive drive;
+    private ArmTeleOp arm;
+
+    @Override
+    public void runOpMode() {
+        drive = new MecanumDrive(hardwareMap, new Pose2d(0, 0, 0));
+
+        initializeHardware();
+
+        if (armMotor != null) {
+            arm = new ArmTeleOp(gamepad2, armMotor, dumperServo);
+            resetDumper();
+        }
+
+        waitForStart();
+
+        while (opModeIsActive()) {
+            resetIMUIfNeeded();
+            driveUsingControllers(false);
+
+            drive.updatePoseEstimate();
+
+            telemetry.addData("x", drive.pose.position.x);
+            telemetry.addData("y", drive.pose.position.y);
+            telemetry.addData("heading", drive.pose.heading);
+
+            // Code added to draw the pose, remove before competition, causes lags:
+            TelemetryPacket p = new TelemetryPacket();
+            Canvas c = p.fieldOverlay();
+            c.setStroke("#3F5100");
+            MecanumDrive.drawRobot(c, drive.pose);
+
+            FtcDashboard dashboard = FtcDashboard.getInstance();
+            dashboard.sendTelemetryPacket(p);
+
+            if (armMotor != null && dumperServo != null && gateServo != null) {
+                outputUsingControllers();
+                intakeUsingControllers();
+                telemetry.addData("ArmMotor", armMotor.getCurrentPosition());
+                telemetry.addData("DumperServo", dumperServo.getPosition());
+                telemetry.addData("GateServo", gateServo.getPosition());
+            }
+
+            telemetry.addData("FRMotor", FR.getCurrentPosition());
+            telemetry.addData("FRMotor", FR.getPowerFloat());
+            telemetry.addData("FLMotor", FL.getCurrentPosition());
+            telemetry.addData("FLMotor", FL.getPowerFloat());
+            telemetry.addData("BRMotor", BR.getCurrentPosition());
+            telemetry.addData("BRMotor", BR.getPowerFloat());
+            telemetry.addData("BLMotor", BL.getCurrentPosition());
+            telemetry.addData("BLMotor", BL.getPowerFloat());
+            telemetry.update();
+        }
+
+        // Close drive (release resources)
+        drive.close();
+    }
+
+    boolean leftBumperIsPressed = false;
+
+    public void resetIMUIfNeeded() {
+        if (gamepad1.left_bumper && !leftBumperIsPressed) {
+            IMU.Parameters parameters;
+            if (drive.isDevBot) {
+                parameters = new IMU.Parameters(new RevHubOrientationOnRobot(
+                        RevHubOrientationOnRobot.LogoFacingDirection.UP,
+                        RevHubOrientationOnRobot.UsbFacingDirection.FORWARD));
+            } else {
+                parameters = new IMU.Parameters(new RevHubOrientationOnRobot(
+                        RevHubOrientationOnRobot.LogoFacingDirection.RIGHT,
+                        RevHubOrientationOnRobot.UsbFacingDirection.UP)); }
+            drive.imu.initialize(parameters);
+        }
+        leftBumperIsPressed = gamepad1.left_bumper;
+    }
+
```

Comment:
As mentioned above, this routine should probably be removed. Even barring that, if it is kept, it needs: 1) a really good explanation of what it does, how to invoke it, when to invoke it, etc. and 2) it needs to be common code that is shared with MecanumDrive.java. This code was copied from MecanumDrive.java with no note that warns that it's been duplicated. This is bad because a bug fix might be done in MecanumDrive.java (maybe one of the directions is wrong) but the fixer wouldn't know that it *also* has to be fixed here. In cases like this, there should be a common function that *both* pieces of code call. But even better would be to remove this function.